File: clients/reactive-dart/lib/src/optimism.dart
import 'dart:async';

import 'contracts.dart';

abstract interface class LocalDurableWrite<T, L> {
  Future<L> commitLocalAndQueue(T value);
  Future<void> commitAuthoritative(T value);
}

abstract interface class RemoteConfirmedWriter<T> {
  Future<T> write(T value);
}

abstract interface class ForegroundSyncCycle<S> {
  void hint();
  Future<S> syncNow();
}

sealed class OptimisticWriteResult<T, L, S> {
  const OptimisticWriteResult({required this.partition});
  final String partition;
}

final class RemoteConfirmedWrite<T, L, S>
    extends OptimisticWriteResult<T, L, S> {
  const RemoteConfirmedWrite({
    required super.partition,
    required this.authoritative,
  });

  final T authoritative;
}

final class LocalQueuedWrite<T, L, S> extends OptimisticWriteResult<T, L, S> {
  const LocalQueuedWrite({
    required super.partition,
    required this.localResult,
  });

  final L localResult;
}

final class LocalConfirmedWrite<T, L, S>
    extends OptimisticWriteResult<T, L, S> {
  const LocalConfirmedWrite({
    required super.partition,
    required this.localResult,
    required this.syncResult,
  });

  final L localResult;
  final S syncResult;
}

Future<OptimisticWriteResult<T, L, S>> writeWithOptimism<T, L, S>({
  required SyncOptimism strategy,
  required SyncSession session,
  required T value,
  required LocalDurableWrite<T, L> local,
  required RemoteConfirmedWriter<T> remote,
  required ForegroundSyncCycle<S> sync,
  FutureOr<void> Function()? wakeBackground,
}) async {
  final identity = requireAuthenticated(session);
  final partition = storagePartitionKey(identity);

  switch (strategy) {
    case SyncOptimism.remoteConfirmed:
      final authoritative = await remote.write(value);
      await local.commitAuthoritative(authoritative);
      return RemoteConfirmedWrite<T, L, S>(
        partition: partition,
        authoritative: authoritative,
      );
    case SyncOptimism.localDurable:
      final localResult = await local.commitLocalAndQueue(value);
      sync.hint();
      await wakeBackground?.call();
      return LocalQueuedWrite<T, L, S>(
        partition: partition,
        localResult: localResult,
      );
    case SyncOptimism.localThenRemote:
      final localResult = await local.commitLocalAndQueue(value);
      sync.hint();
      await wakeBackground?.call();
      final syncResult = await sync.syncNow();
      return LocalConfirmedWrite<T, L, S>(
        partition: partition,
        localResult: localResult,
        syncResult: syncResult,
      );
  }
}

File: clients/reactive-ts/src/optimism.ts
import type { ConsistencyPolicyId, SyncOptimism, SyncSession } from './contracts.ts';
import {
  SYNC_OPTIMISM,
  requireAuthenticated,
  resolveSyncOptimism,
  storagePartitionKey,
} from './contracts.ts';

export interface LocalDurableWrite<T, R = unknown> {
  /** Commit the app projection and durable opto-sync queue entry atomically. */
  commitLocalAndQueue(value: T): Promise<R>;
  /** Persist a server-confirmed value without creating another pending write. */
  commitAuthoritative(value: T): Promise<void>;
}

export interface RemoteConfirmedWriter<T> {
  write(value: T, signal: AbortSignal): Promise<T>;
}

export interface ForegroundSyncCycle<R = unknown> {
  syncNow(): Promise<R>;
  hint?(): void;
}

export interface OptimisticWriteOptions<T, L = unknown, S = unknown> {
  strategy: SyncOptimism | ConsistencyPolicyId | string;
  session: SyncSession;
  value: T;
  local: LocalDurableWrite<T, L>;
  remote: RemoteConfirmedWriter<T>;
  sync: ForegroundSyncCycle<S>;
  signal?: AbortSignal;
  wakeBackground?: () => void | Promise<void>;
}

export type OptimisticWriteResult<T, L, S> =
  | {
      strategy: 'remote-confirmed';
      status: 'confirmed';
      partition: string;
      authoritative: T;
    }
  | {
      strategy: 'local-durable';
      status: 'queued';
      partition: string;
      localResult: L;
    }
  | {
      strategy: 'local-then-remote';
      status: 'confirmed';
      partition: string;
      localResult: L;
      syncResult: S;
    };

/**
 * Execute one declared optimism level instead of hiding latency/durability
 * trade-offs in framework-specific UI code.
 */
export async function writeWithOptimism<T, L = unknown, S = unknown>(
  options: OptimisticWriteOptions<T, L, S>,
): Promise<OptimisticWriteResult<T, L, S>> {
  const identity = requireAuthenticated(options.session);
  const partition = storagePartitionKey(identity);
  const signal = options.signal ?? new AbortController().signal;
  const strategy = resolveSyncOptimism(options.strategy);

  switch (strategy) {
    case SYNC_OPTIMISM.remoteConfirmed: {
      // Pessimistic/online-only: do not mutate local state until the server has
      // durably accepted and returned the authoritative representation.
      const authoritative = await options.remote.write(options.value, signal);
      await options.local.commitAuthoritative(authoritative);
      return {
        strategy: SYNC_OPTIMISM.remoteConfirmed,
        status: 'confirmed',
        partition,
        authoritative,
      };
    }

    case SYNC_OPTIMISM.localDurable: {
      // Offline-first: the caller's transaction must write the visible row and
      // queue entry together. Background wake-up is only a hint; durability is
      // already complete when this promise resolves.
      const localResult = await options.local.commitLocalAndQueue(options.value);
      options.sync.hint?.();
      await options.wakeBackground?.();
      return {
        strategy: SYNC_OPTIMISM.localDurable,
        status: 'queued',
        partition,
        localResult,
      };
    }

    case SYNC_OPTIMISM.localThenRemote: {
      // Optimistic UI plus a completion boundary for workflows that may render
      // immediately but cannot leave the screen until the protocol cycle lands.
      const localResult = await options.local.commitLocalAndQueue(options.value);
      options.sync.hint?.();
      await options.wakeBackground?.();
      const syncResult = await options.sync.syncNow();
      return {
        strategy: SYNC_OPTIMISM.localThenRemote,
        status: 'confirmed',
        partition,
        localResult,
        syncResult,
      };
    }
  }
}
